include/swift/AST/Decl.h

```diff
@@ -6478,9 +6482,18 @@ class AbstractFunctionDecl : public GenericContext, public ValueDecl {
   /// \sa hasBody()
   BraceStmt *getBody(bool canSynthesize = true) const;
 
+  struct TypecheckedBodyResult {
+    /// The body of the function with as many statements type checked as
+    /// possible.
+    BraceStmt *TypecheckedBody;
+    /// If a single statement in the type checked body has an error, this is
+    /// \c true.
+    bool HadError;
+  };
+
   /// Retrieve the type-checked body of the given function, or \c nullptr if
   /// there's no body available.
-  BraceStmt *getTypecheckedBody() const;
+  TypecheckedBodyResult getTypecheckedBody() const;
```

Comment:
No, because editor functionality (most notably cursor info + indexing) still wants to walk the partially type-checked brace statement to salvage as much information as possible from it. If we just return `ErrorExpr` (or similar here), that’s no longer possible.

Reply:
Hmm okay, makes sense. Eventually we really ought to do this at a finer grained level though, e.g have `typeCheckExpression` produce individual `ErrorExpr`s or type expressions as `ErrorType`. It's really unfortunate to have type-checking requests return un-type-checked AST on failure.